x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/final_readonly/fields/data_source/data_source.stories.tsx

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import type { Story } from '@storybook/react';
+import { FieldReadOnly } from '../../field_readonly';
+import type { DiffableAllFields } from '../../../../../../../../../common/api/detection_engine';
+import { ThreeWayDiffStorybookProviders } from '../../storybook/three_way_diff_storybook_providers';
+import {
+  dataSourceWithDataView,
+  dataSourceWithIndexPatterns,
+  mockDataView,
+} from '../../storybook/mocks';
+
+export default {
+  component: FieldReadOnly,
+  title: 'Rule Management/Prebuilt Rules/Upgrade Flyout/ThreeWayDiff/FieldReadOnly/data_source',
+};
+
+interface TemplateProps {
+  finalDiffableRule: Partial<DiffableAllFields>;
+  kibanaServicesMock?: Record<string, unknown>;
+}
+
+const Template: Story<TemplateProps> = (args) => {
+  return (
+    <ThreeWayDiffStorybookProviders kibanaServicesMock={args.kibanaServicesMock}>
+      <FieldReadOnly
+        fieldName="data_source"
+        finalDiffableRule={args.finalDiffableRule as DiffableAllFields}
```

Comment:
I'm casting it here to be able to pass partial mocks into `FieldReadOnly`. I have to use `DiffableAllFields` as a prop type for `finalDiffableRule` because it's the only type that would let me render different field components based on `fieldName`.

IMO, it's fine for Storybooks, but I agree it's not ideal for prod code. When we start integrating context with `FieldReadOnly` we'll probably have to come up with a different type. Do you have an idea of how this can be done? Let's discuss offline.